### PR TITLE
Rework WiFiClient

### PR DESCRIPTION
**Pull request by schrod**
Rework WiFiClient to correct error where making a copy of a WiFiClient object resulted in the socket being closed prematurely.

Added loop and select to write to handle/prevent EAGAIN errors.

### CODE REVIEW

**Comment by me-no-dev** [comment]
There are generally two ways to tackle issues like that. I would rather use shared pointer which will manage the state on it's own, will clean the code and you will not have to worry to ref/unref count.
Please do have ```Added loop and select to write to handle/prevent EAGAIN errors.``` in a separate PR as the change has no relation to the rest.

**Comment by schrod** [comment]
Interesting, I hadn't been aware of shared pointers (I'm mostly a C developer). I agree that looks like it will be a cleaner implementation, I'll work on that change and move the changes to write to a separate PR.

File: libraries/WiFi/src/WiFiClient.h [hunk 1]
```diff
@@ -20,30 +20,21 @@
 #ifndef _WIFICLIENT_H_
 #define _WIFICLIENT_H_
 
+
 #include "Arduino.h"
 #include "Client.h"
+#include <memory>
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiClient.h]: You can move the whole class in the CPP and only forward declare it here. It's private really and users do not need to bother with it.

File: libraries/WiFi/src/WiFiClient.cpp [hunk 2]
```diff
@@ -199,52 +165,21 @@ int WiFiClient::read()
 
 size_t WiFiClient::write(const uint8_t *buf, size_t size)
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiClient.cpp]: You need to sync this code now with the one that I already merged :)

File: libraries/WiFi/src/WiFiClient.h [hunk 3]
```diff
@@ -69,21 +87,26 @@ class WiFiClient : public Client
         return !this->operator==(rhs);
     };
 
-    int fd()
+    int fd() const
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiClient.h]: this should go into the CPP

**Comment by me-no-dev** [comment]
great work :) maybe you want to look at WiFiClientSecure?

**Comment by schrod** [comment]
Yup, looks like WiFiClientSecure has the same issue. I'll see what I can do.

**Comment by me-no-dev** [comment]
udp probably suffers from that same thing :D

**Pull request merged**